source/Octopus.Tentacle/Kubernetes/Crypto/ScriptPodLogEncryptionKeyProvider.cs

```diff
@@ -0,0 +1,109 @@
+    {
+        const int KeyLengthInBytes = 32;
+        const string Filename = "keyfile";
+        readonly IScriptWorkspaceFactory scriptWorkspaceFactory;
+        readonly IScriptPodLogEncryptionKeyGenerator encryptionKeyGenerator;
+
+        readonly ConcurrentDictionary<ScriptTicket, byte[]> encryptionKeyCache = new();
+
+        public ScriptPodLogEncryptionKeyProvider(IScriptWorkspaceFactory scriptWorkspaceFactory, IScriptPodLogEncryptionKeyGenerator encryptionKeyGenerator)
+        {
+            this.scriptWorkspaceFactory = scriptWorkspaceFactory;
+            this.encryptionKeyGenerator = encryptionKeyGenerator;
+        }
+
+        public async Task WriteEncryptionKeyfileToWorkspace(ScriptTicket scriptTicket, CancellationToken cancellationToken)
+        {
+            if (encryptionKeyCache.ContainsKey(scriptTicket))
+            {
+                throw new PodLogEncryptionKeyException($"An encryption key already exists for script {scriptTicket.TaskId}");
+            }
+            
+            var workspace = scriptWorkspaceFactory.GetWorkspace(scriptTicket);
+            await GenerateAndWriteEncryptionKeyfileToWorkspace(scriptTicket, workspace, cancellationToken);
+        }
+
+        async Task<byte[]> GenerateAndWriteEncryptionKeyfileToWorkspace(ScriptTicket scriptTicket, IScriptWorkspace workspace,CancellationToken cancellationToken)
+        {
+            var encryptionKeyBytes = await encryptionKeyGenerator.GenerateKey(scriptTicket, KeyLengthInBytes, cancellationToken);
+            if (!encryptionKeyCache.TryAdd(scriptTicket, encryptionKeyBytes))
+            {
+                throw new PodLogEncryptionKeyException($"Failed to store encryption key in memory cache for script {scriptTicket.TaskId}");
+            }
+
+            try
+            {
+                var fileContents = Convert.ToBase64String(encryptionKeyBytes);
+                workspace.WriteFile(Filename, fileContents);
+                return encryptionKeyBytes;
+            }
+            catch (Exception e)
+            {
+                throw new PodLogEncryptionKeyException($"Failed to write encryption key to workspace for script {scriptTicket.TaskId}", e);
+            }
+        }
+
+        public async Task<byte[]> GetEncryptionKey(ScriptTicket scriptTicket, CancellationToken cancellationToken)
+        {
+            if (encryptionKeyCache.TryGetValue(scriptTicket, out var keyBytes))
+            {
+                return keyBytes;
+            }
+
+            var workspace = scriptWorkspaceFactory.GetWorkspace(scriptTicket);
+            var fileContents = workspace.TryReadFile(Filename);
+            //If we can't load the encryption key from the filesystem
+            if (fileContents == null)
+            {
+                //regenerate the encryption key, write to the filesystem and return the key
+                return await GenerateAndWriteEncryptionKeyfileToWorkspace(scriptTicket, workspace, cancellationToken);
+            }
```

Comment:
When tentacle gets the encryption key, if it can't load it from memory or the script workspace, we regenerate the same key and store it in memory and the disk. This should only happen _after_ a tentacle & NFS pod restart.